Source/HelixToolkit.Wpf.Shared/Helpers/Viewport3DHelper.cs

```diff
@@ -223,48 +223,65 @@ public static IEnumerable<RectangleHitResult> FindHits(this Viewport3D viewport,
-                    for (var i = 0; i < geometry.TriangleIndices.Count / 3; i++)
+                    if (rectangle.Contains(visualBound)) // object is both inside and touch
+                    {
+                        results.Add(new RectangleHitResult(model, visual));
+                    }
+                    else if (rectangle.IntersectsWith(visualBound))
                     {
-                        var triangle = new Triangle(
-                            point2Ds[geometry.TriangleIndices[i * 3]],
-                            point2Ds[geometry.TriangleIndices[(i * 3) + 1]],
-                            point2Ds[geometry.TriangleIndices[(i * 3) + 2]]);
-                        switch (mode)
+                        // transform the positions of the mesh to screen coordinates
+                        var point2Ds = geometry.Positions.Select(transform.Transform).Select(viewport.Point3DtoPoint2D).ToArray();
+                        var meshPoints = new Point[geometry.TriangleIndices.Count];
+                        for (var i = 0; i < geometry.TriangleIndices.Count; i += 3)
                         {
-                            case SelectionHitMode.Inside:
-                                status = status && triangle.IsCompletelyInside(rectangle);
-                                break;
-                            case SelectionHitMode.Touch:
-                                status = status
-                                         || triangle.IsCompletelyInside(rectangle)
-                                         || triangle.IntersectsWith(rectangle)
-                                         || triangle.IsRectCompletelyInside(rectangle);
-                                break;
+                            meshPoints[i] = point2Ds[geometry.TriangleIndices[i]];
+                            meshPoints[i + 1] = point2Ds[geometry.TriangleIndices[i + 1]];
+                            meshPoints[i + 2] = point2Ds[geometry.TriangleIndices[i + 2]];
                         }
+                        double minX = meshPoints.Min(x => x.X);
+                        double minY = meshPoints.Min(x => x.Y);
+                        double maxX = meshPoints.Max(x => x.X);
+                        double maxY = meshPoints.Max(x => x.Y);
+                        Rect geometryBound = new Rect(new Point(minX, minY), new Point(maxX, maxY));
 
-                        if (mode == SelectionHitMode.Touch && status)
+                        if (rectangle.Contains(geometryBound))// object is both inside and touch
                         {
-                            break;
+                            results.Add(new RectangleHitResult(model, visual));
+                        }
+                        else if (mode == SelectionHitMode.Touch && rectangle.IntersectsWith(geometryBound))
+                        {
+                            // evaluate each triangle
+                            for (var i = 0; i < meshPoints.Length; i += 3)
```

Comment:
Here as well, only need to evaluate whether any vertex is inside the rectangle.

Reply:
Instead of transform all triangles, just transform all vertices, then use triangle indices to get the transformed vertices back during the intersection testing.

Reply:
`meshPositions` should have count = `geometry.Positions.count`, otherwise you are still duplicating vertex transforms based on triangle indices. For example, if two triangle has indices of `1, 2, 3` and `1, 2, 4`, now you are transforming 1 and 2 vertices twice.